kata 2.0: delete use_vsock option and proxy abstraction

Pennyzct opened the pull request:
# Description of problem
with kata containers moving to 2.0, the default qemu type still uses _serial port plus kata-proxy_ to communicate to agent, which leaves the following error: 
```
$ sudo ctr run --runtime io.containerd.run.kata.v2 -t --rm docker.io/library/busybox:latest hello sh
ctr: Failed to check if grpc server is working: context deadline exceeded: unknown
```

So, We need to deprecate serial port and related parts, like yamux, kata-proxy, etc. `kata proxy` as additional component used to handle multiplexing on serial port.

Besides Cleaning up related unit tests, I'd also import a new socket type `MockHybridVSock` to deal with new ttrpc-based hybrid-vsock mock server.

Fixes: #389

Signed-off-by: Penny Zheng penny.zheng@arm.com

Pennyzct commented:
/test

Pennyzct commented:
/retest

bergwolf commented:
@Pennyzct Need rebase /-.-/

Pennyzct commented:
@bergwolf sorry for the delay~~ caught me quite a time to rebase. 😐

bergwolf commented:
@Pennyzct Sorry this is conflicting again ;-(
```
Conflicting files
src/runtime/virtcontainers/api_test.go
```

liubin commented on src/runtime/containerd-shim-v2/monitor_address:
```diff
@@ -0,0 +1 @@
+/containerd-shim/UnitTest/777-77-77777777/shim-monitor.sock
```
I'm sorry that this sock file should not be commit into repo.

liubin commented on src/runtime/virtcontainers/api.go:
```diff
@@ -195,16 +195,6 @@ func FetchSandbox(ctx context.Context, sandboxID string) (VCSandbox, error) {
-	// If the agent is long live connection, it needs to restart the proxy to
-	// watch the guest console if it hadn't been watched.
-	if s.agent.longLiveConn() {
-		err = s.startProxy()
```
May this lose guest console output ?

Pennyzct replied:
yes, it will be. previously, when we use vsock as connection socket, different hypervisors use different methods to deliver guest console output, that is, qemu uses shim, firecracker uses another goroutine, etc.

liubin commented on src/runtime/virtcontainers/proxy.go:
```diff
@@ -1,270 +0,0 @@
-	// For firecracker, it hasn't support the console watching and it's consoleURL
-	// will be set empty.
-	// TODO: add support for hybrid vsocks, see https://github.com/kata-containers/runtime/issues/2098
-	if params.debug && params.consoleURL != "" && !strings.HasPrefix(params.consoleURL, kataclient.HybridVSockScheme) {
```
We should keep watching console and use something like `params.debug` to control it.

bpradipt replied:
Yeah makes sense.. console is really useful for debugging.

Pennyzct commented:
@liubin @bergwolf 
We could use fc current logging method to deliver  the guest console output.
```
	if fc.config.Debug {
		stdin, err := fc.watchConsole()
```
see details here:
https://github.com/kata-containers/kata-containers/blob/2bff7a16f5e2499a7f9c159d1350ef8b5c87c9ad/src/runtime/virtcontainers/fc.go#L1221

bergwolf commented:
@Pennyzct Yes, and we need the same functionality for other hypervisors as well.

Pennyzct commented:
Hi @bergwolf @liubin sorry for the delay response. ;)
Addressing your suggestions, I just raised a new commit to deliver guest console output for all hypervisors. 
```
console-watcher: use console watcher to monitor guest console outputs
Import new console watcher to monitor guest console outputs, and will be
only effective when we turn on enable_debug option.
Guest console outputs may include guest kernel debug info, agent debug info,
etc.
```
And another small commit to fix launching error when running with firecrakcer.
Plz review~~~thx!!!

bergwolf commented:
/test

c3d commented on src/runtime/cli/config/configuration-qemu-virtiofs.toml.in:
```diff
@@ -222,12 +221,6 @@ vhost_user_store_path = "@DEFVHOSTUSERSTOREPATH@"
-# with the agent and no proxy is started, otherwise use unix
-# sockets and start a proxy to communicate with the agent.
-# Default false
-#use_vsock = true
```
Shouldn't we default to "true" and warn if the option is set to false, instead of rejecting the option entirely?

Alternatively, start a document that lists all options that are removed from 2.0 so that people can proactively edit their configuration files.

bergwolf replied:
We are dropping serial port support in 2.0, -- that's why `use_vsock` is just removed instead of switching to `true` as default.
